### PR TITLE
basic nnbd awareness for `avoid_init_to_null`

### PR DESCRIPTION
**Pull request by pq**
First cut.

Lint description: http://dart-lang.github.io/linter/lints/avoid_init_to_null.html

(Used in `pedantic` and `flutter`.)

Test ideas and suggestions welcome!

See: #1770 

/cc @stereotype441 @scheglov 

/fyi @munificent

### CODE REVIEW

File: lib/src/rules/avoid_init_to_null.dart [hunk 1]
```diff
@@ -66,21 +68,30 @@ class AvoidInitToNull extends LintRule implements NodeLintRule {
...
 }
 
 class _Visitor extends SimpleAstVisitor<void> {
   final LintRule rule;
+  final LinterContext context;
 
-  _Visitor(this.rule);
+  final bool nnbdEnabled;
+  _Visitor(this.rule, this.context)
+      : nnbdEnabled = context.analysisOptions.contextFeatures
```
Comment (scheglov) [inline lib/src/rules/avoid_init_to_null.dart]: This is not fully correct.
Individual libraries might specify their language versions, so `FeatureSet`.
The feature set from the context is the default value when not overridden on the library level.

So, ask the unit for its `CompilationUnit.featureSet` instead.

Comment (stereotype441) [inline lib/src/rules/avoid_init_to_null.dart]: It's possible for NNBD to be enabled in some files and not others.  The right way to check for this is using CompilationUnit.featureSet or LibraryElement.isNonNullableByDefault

File: lib/src/rules/avoid_init_to_null.dart [hunk 2]
```diff
@@ -89,7 +100,9 @@ class _Visitor extends SimpleAstVisitor<void> {
     if (!node.isConst &&
         !node.isFinal &&
         DartTypeUtilities.isNullLiteral(node.initializer)) {
-      rule.reportLint(node);
+      if (isNullable(node.declaredElement.type)) {
```
Comment (scheglov) [inline lib/src/rules/avoid_init_to_null.dart]: The enclosing `if` combines multiple checks with `&&`.
So, probably this additional check could be added there.

Same for the similar check for parameters above.

File: test/rules/experiments/nnbd/rules/avoid_init_to_null.dart [hunk 3]
```diff
@@ -18,12 +24,20 @@ class X {
...
   // TODO(pq): ints are not nullable so we'll want to update the lint here
   // since it will produce a compilation error.
-  // int x = null; //LINT
+  int x = null; //OK (compilation error)
+  int? xx = null; //LINT
   int y; //OK
-  int z;
+  int z; //OK
+
+  X({int a: null}) //OK (compilation error)
```
Comment (scheglov) [inline test/rules/experiments/nnbd/rules/avoid_init_to_null.dart]: Test for nullable parameter?

File: lib/src/rules/avoid_init_to_null.dart (hunk 1 above)
Comment (pq) [reply lib/src/rules/avoid_init_to_null.dart]: super thanks!

Comment (pq) [reply lib/src/rules/avoid_init_to_null.dart]: excellent.  Paul pointed this out too.  Thank you!

File: test/rules/experiments/nnbd/rules/avoid_init_to_null.dart (hunk 3 above)
Comment (pq) [reply test/rules/experiments/nnbd/rules/avoid_init_to_null.dart]: See `{this.xx: null}` below.  Is that what you mean?

Comment (scheglov) [reply test/rules/experiments/nnbd/rules/avoid_init_to_null.dart]: Ah, you reused fields to also test parameters.
Clever! :-)

**Pull request merged**